File: src/commands/pr/create.ts
import chalk from 'chalk';
import ora from 'ora';
import inquirer from 'inquirer';
import { loadManifest, getAllRepoInfo, loadState, saveState, parseGitHubUrl, getManifestRepoInfo } from '../../lib/manifest.js';
import {
  pathExists,
  getCurrentBranch,
  hasCommitsAhead,
  pushBranch,
  remoteBranchExists,
  isGitRepo,
} from '../../lib/git.js';
import {
  createPullRequest,
  createLinkedPRs,
  generateManifestPRBody,
  findPRByBranch,
} from '../../lib/github.js';
import { linkBranchToManifestPR, saveLinkedPRs } from '../../lib/linker.js';
import type { RepoInfo, PRCreateOptions, LinkedPR } from '../../types.js';

interface CreateOptions {
  title?: string;
  body?: string;
  draft?: boolean;
  base?: string;
  push?: boolean;
}

/**
 * Create linked PRs across all repositories with changes
 */
export async function createPR(options: CreateOptions = {}): Promise<void> {
  const { manifest, rootDir } = await loadManifest();
  const repos = getAllRepoInfo(manifest, rootDir);

  // Check which repos are cloned
  const clonedRepos: RepoInfo[] = [];
  for (const repo of repos) {
    if (await pathExists(repo.absolutePath)) {
      clonedRepos.push(repo);
    }
  }

  if (clonedRepos.length === 0) {
    console.log(chalk.yellow('No repositories are cloned.'));
    return;
  }

  // Get current branch and check for changes in each repo
  const repoStatus = await Promise.all(
    clonedRepos.map(async (repo) => {
      const branch = await getCurrentBranch(repo.absolutePath);
      const hasChanges = await hasCommitsAhead(repo.absolutePath, repo.default_branch);
      const needsPush = hasChanges && !(await remoteBranchExists(repo.absolutePath, branch));
      return { repo, branch, hasChanges, needsPush };
    })
  );

  // Check manifest for changes too
  const manifestInfo = getManifestRepoInfo(manifest, rootDir);
  let manifestBranch: string | null = null;
  let manifestHasChanges = false;
  let manifestNeedsPush = false;
  if (manifestInfo && await isGitRepo(manifestInfo.absolutePath)) {
    manifestBranch = await getCurrentBranch(manifestInfo.absolutePath);
    manifestHasChanges = await hasCommitsAhead(manifestInfo.absolutePath, manifestInfo.default_branch);
    manifestNeedsPush = manifestHasChanges && !(await remoteBranchExists(manifestInfo.absolutePath, manifestBranch));
  }

  // Filter to repos with changes
  const withChanges = repoStatus.filter((r) => r.hasChanges);

  if (withChanges.length === 0 && !manifestHasChanges) {
    console.log(chalk.yellow('No repositories have commits ahead of their default branch.'));
    console.log(chalk.dim('Make some commits first, then run this command again.'));
    return;
  }

  // Only check branch consistency for repos WITH CHANGES (not all repos)
  const branchesWithChanges = withChanges.map((r) => r.branch);
  if (manifestHasChanges && manifestBranch) {
    branchesWithChanges.push(manifestBranch);
  }
  const uniqueBranches = [...new Set(branchesWithChanges)];

  if (uniqueBranches.length > 1) {
    console.log(chalk.yellow('Repositories with changes are on different branches:'));
    for (const { repo, branch } of withChanges) {
      console.log(`  ${repo.name}: ${chalk.cyan(branch)}`);
    }
    if (manifestHasChanges && manifestInfo && manifestBranch) {
      console.log(`  ${manifestInfo.name}: ${chalk.cyan(manifestBranch)}`);
    }
    console.log('');
    console.log(chalk.dim('Use `gitgrip checkout <branch>` to sync branches first.'));
    return;
  }

  const branchName = uniqueBranches[0];

  // Check it's not the default branch
  const onDefaultBranch = withChanges.some((r) => r.repo.default_branch === branchName);
  if (onDefaultBranch) {
    console.log(chalk.yellow(`You're on the default branch (${branchName}).`));
    console.log(chalk.dim('Create a feature branch first with `gitgrip branch <name>`.'));
    return;
  }

  console.log(chalk.blue(`Creating PRs for branch: ${chalk.cyan(branchName)}\n`));

  const totalChanges = withChanges.length + (manifestHasChanges ? 1 : 0);
  console.log(`Found changes in ${totalChanges} repos:`);
  for (const { repo } of withChanges) {
    console.log(`  ${chalk.green('•')} ${repo.name}`);
  }
  if (manifestHasChanges && manifestInfo) {
    console.log(`  ${chalk.green('•')} ${manifestInfo.name}`);
  }
  console.log('');

  // Check if any need to be pushed first (including manifest)
  const needsPush = withChanges.filter((r) => r.needsPush);
  const allNeedsPush: { repo: RepoInfo; needsPush: boolean }[] = [...needsPush];
  if (manifestNeedsPush && manifestInfo) {
    allNeedsPush.push({ repo: manifestInfo, needsPush: true });
  }

  if (allNeedsPush.length > 0) {
    if (options.push) {
      console.log(chalk.dim('Pushing branches to remote...\n'));
      for (const { repo } of allNeedsPush) {
        const spinner = ora(`Pushing ${repo.name}...`).start();
        try {
          await pushBranch(repo.absolutePath, branchName, 'origin', true);
          spinner.succeed(`${repo.name}: pushed`);
        } catch (error) {
          spinner.fail(`${repo.name}: ${error instanceof Error ? error.message : error}`);
          console.log(chalk.red('\nFailed to push. Fix the error and try again.'));
          return;
        }
      }
      console.log('');
    } else {
      console.log(chalk.yellow('Some branches need to be pushed to remote first:'));
      for (const { repo } of allNeedsPush) {
        console.log(`  ${repo.name}`);
      }
      console.log('');
      console.log(chalk.dim('Run with --push flag to push automatically, or push manually.'));
      return;
    }
  }

  // Get PR title if not provided
  let title: string = options.title ?? '';
  if (!title) {
    const answers = await inquirer.prompt([
      {
        type: 'input',
        name: 'title',
        message: 'PR title:',
        default: branchName.replace(/[-_]/g, ' ').replace(/^feature\//, ''),
        validate: (input: string) => input.length > 0 || 'Title is required',
      },
    ]);
    title = answers.title as string;
  }

  // Get PR body if not provided
  let body = options.body ?? '';
  if (!body) {
    const answers = await inquirer.prompt([
      {
        type: 'editor',
        name: 'body',
        message: 'PR description (optional):',
        default: '',
      },
    ]);
    body = answers.body.trim();
  }

  // Create PRs
  const spinner = ora('Creating pull requests...').start();

  try {
    const reposForPR = withChanges.map((r) => r.repo);
    const prOptions: PRCreateOptions = {
      title,
      body,
      draft: options.draft,
      base: options.base,
    };

    // Create PRs in each repo
    const linkedPRs = await createLinkedPRs(reposForPR, branchName, prOptions);

    // Create manifest PR if manifest has changes
    let manifestPR: LinkedPR | null = null;
    if (manifestHasChanges && manifestInfo) {
      try {
        const manifestPRResult = await createPullRequest(
          manifestInfo.owner,
          manifestInfo.repo,
          branchName,
          manifestInfo.default_branch,
          title,
          body,
          options.draft
        );
        manifestPR = {
          repoName: manifestInfo.name,
          owner: manifestInfo.owner,
          repo: manifestInfo.repo,
          number: manifestPRResult.number,
          url: manifestPRResult.url,
          state: 'open',
          approved: false,
          checksPass: true,
          mergeable: true,
        };
      } catch (error) {
        // Don't fail the whole operation if manifest PR fails
        console.log(chalk.yellow(`\nWarning: Could not create manifest PR: ${error instanceof Error ? error.message : error}`));
      }
    }

    spinner.succeed('Pull requests created!\n');

    // Display results
    console.log(chalk.green('Created PRs:'));
    for (const pr of linkedPRs) {
      console.log(`  ${pr.repoName}: ${chalk.cyan(pr.url)}`);
    }
    if (manifestPR) {
      console.log(`  ${manifestPR.repoName}: ${chalk.cyan(manifestPR.url)}`);
    }

    // Generate a summary for the user
    console.log('');
    console.log(chalk.dim('To view PR status: gitgrip pr status'));
    console.log(chalk.dim('To merge all PRs:  gitgrip pr merge'));

    // Save state
    const state = await loadState(rootDir);
    // We don't have a manifest PR number in simple mode, use branch name as key
    state.branchToPR[branchName] = manifestPR?.number ?? -1;
    await saveState(rootDir, state);
  } catch (error) {
    spinner.fail('Failed to create PRs');
    console.error(chalk.red(error instanceof Error ? error.message : String(error)));
  }
}

File: src/cli/commands/pr/create.rs
//! PR create command implementation

use crate::cli::output::Output;
use crate::core::manifest::{Manifest, PlatformType};
use crate::core::repo::RepoInfo;
use crate::core::state::StateFile;
use crate::git::{get_current_branch, open_repo, path_exists};
use crate::platform::{detect_platform, get_platform_adapter};
use git2::Repository;
use std::path::PathBuf;

/// Run the PR create command
pub async fn run_pr_create(
    workspace_root: &PathBuf,
    manifest: &Manifest,
    title: Option<&str>,
    body: Option<&str>,
    draft: bool,
    push_first: bool,
    dry_run: bool,
) -> anyhow::Result<()> {
    if dry_run {
        Output::header("PR Preview");
        println!();
    } else {
        Output::header("Creating pull requests...");
        println!();
    }

    let repos: Vec<RepoInfo> = manifest
        .repos
        .iter()
        .filter_map(|(name, config)| RepoInfo::from_config(name, config, workspace_root))
        .filter(|r| !r.reference) // Skip reference repos
        .collect();

    // Get current branch for all repos and verify consistency
    let mut branch_name: Option<String> = None;
    let mut repos_with_changes: Vec<RepoInfo> = Vec::new();

    for repo in &repos {
        if !path_exists(&repo.absolute_path) {
            continue;
        }

        match open_repo(&repo.absolute_path) {
            Ok(git_repo) => {
                let current = match get_current_branch(&git_repo) {
                    Ok(b) => b,
                    Err(_) => continue,
                };

                // Skip if on default branch
                if current == repo.default_branch {
                    continue;
                }

                // Check for changes ahead of default branch
                if has_commits_ahead(&git_repo, &current, &repo.default_branch)? {
                    if let Some(ref bn) = branch_name {
                        if bn != &current {
                            anyhow::bail!(
                                "Repositories are on different branches: {} vs {}",
                                bn,
                                current
                            );
                        }
                    } else {
                        branch_name = Some(current);
                    }
                    repos_with_changes.push(repo.clone());
                }
            }
            Err(e) => Output::error(&format!("{}: {}", repo.name, e)),
        }
    }

    // Also check the manifest repo for changes
    if let Some(manifest_config) = &manifest.manifest {
        let manifests_dir = workspace_root.join(".gitgrip").join("manifests");

        // Only process if manifest repo exists and has a git directory
        if !manifests_dir.join(".git").exists() || !path_exists(&manifests_dir) {
            // No manifest git repo found - skip
        } else if let Some(manifest_repo) =
            create_manifest_repo_info(manifest_config, &manifests_dir, workspace_root)
        {
            // Check if manifest repo has changes
            match check_repo_for_changes(&manifest_repo, &mut branch_name) {
                Ok(true) => {
                    repos_with_changes.push(manifest_repo);
                }
                Ok(false) => {
                    // No changes or on default branch - skip
                }
                Err(e) => {
                    Output::warning(&format!("Could not check manifest repo: {}", e));
                }
            }
        }
    }

    let branch = match branch_name {
        Some(b) => b,
        None => {
            println!("No repositories have changes to create PRs for.");
            return Ok(());
        }
    };

    // Get title from argument or use branch name as fallback
    let pr_title = title.map(|s| s.to_string()).unwrap_or_else(|| {
        // Convert branch name to title: feat/my-feature -> My feature
        let title = branch
            .trim_start_matches("feat/")
            .trim_start_matches("fix/")
            .trim_start_matches("chore/")
            .replace(['-', '_'], " ");
        let mut chars = title.chars();
        match chars.next() {
            None => title,
            Some(first) => first.to_uppercase().collect::<String>() + chars.as_str(),
        }
    });

    // Push if requested (skip for preview)
    if push_first && !dry_run {
        Output::info("Pushing branches first...");
        for repo in &repos_with_changes {
            if let Ok(git_repo) = open_repo(&repo.absolute_path) {
                let spinner = Output::spinner(&format!("Pushing {}...", repo.name));
                match crate::git::remote::push_branch(&git_repo, &branch, "origin", true) {
                    Ok(()) => spinner.finish_with_message(format!("{}: pushed", repo.name)),
                    Err(e) => {
                        spinner.finish_with_message(format!("{}: push failed - {}", repo.name, e))
                    }
                }
            }
        }
        println!();
    }

    // Preview mode: show what would be created
    if dry_run {
        Output::info(&format!("Branch: {}", branch));
        Output::info(&format!("Title: {}", pr_title));
        if let Some(pr_body) = body {
            Output::info(&format!("Body: {}", pr_body));
        }
        if draft {
            Output::info("Type: Draft PR");
        }
        println!();

        Output::subheader("Repositories that would create PRs:");
        for repo in &repos_with_changes {
            println!("  - {} ({}/{})", repo.name, repo.owner, repo.repo);
        }
        println!();
        Output::warning("Run without --dry-run to actually create the PRs.");
        return Ok(());
    }

    // Create PRs for each repo
    let mut created_prs: Vec<(String, u64, String)> = Vec::new(); // (repo_name, pr_number, url)

    for repo in &repos_with_changes {
        let platform_type = detect_platform(&repo.url);
        let platform = get_platform_adapter(platform_type, None);

        let spinner = Output::spinner(&format!("Creating PR for {}...", repo.name));

        match platform
            .create_pull_request(
                &repo.owner,
                &repo.repo,
                &branch,
                &repo.default_branch,
                &pr_title,
                body,
                draft,
            )
            .await
        {
            Ok(pr) => {
                spinner.finish_with_message(format!(
                    "{}: created PR #{} - {}",
                    repo.name, pr.number, pr.url
                ));
                created_prs.push((repo.name.clone(), pr.number, pr.url.clone()));
            }
            Err(e) => {
                spinner.finish_with_message(format!("{}: failed - {}", repo.name, e));
            }
        }
    }

    // Save state
    if !created_prs.is_empty() {
        let state_path = workspace_root.join(".gitgrip").join("state.json");
        let mut state = if state_path.exists() {
            let content = std::fs::read_to_string(&state_path)?;
            StateFile::parse(&content).unwrap_or_default()
        } else {
            StateFile::default()
        };

        // Use the first PR number for branch mapping
        if let Some((_, first_pr_number, _)) = created_prs.first() {
            state.set_pr_for_branch(&branch, *first_pr_number);
        }

        let state_json = serde_json::to_string_pretty(&state)?;
        std::fs::write(&state_path, state_json)?;
    }

    // Summary
    println!();
    if created_prs.is_empty() {
        Output::warning("No PRs were created.");
    } else {
        Output::success(&format!("Created {} PR(s):", created_prs.len()));
        for (repo_name, pr_number, url) in &created_prs {
            println!("  {}: #{} - {}", repo_name, pr_number, url);
        }
    }

    Ok(())
}

/// Check if a branch has commits ahead of another branch
fn has_commits_ahead(repo: &Repository, branch: &str, base: &str) -> anyhow::Result<bool> {
    let local_ref = format!("refs/heads/{}", branch);
    let base_ref = format!("refs/remotes/origin/{}", base);

    let local = match repo.find_reference(&local_ref) {
        Ok(r) => r,
        Err(_) => return Ok(false),
    };

    let base_branch = match repo.find_reference(&base_ref) {
        Ok(r) => r,
        Err(_) => {
            // Try local base branch
            match repo.find_reference(&format!("refs/heads/{}", base)) {
                Ok(r) => r,
                Err(_) => return Ok(false),
            }
        }
    };

    let local_oid = local
        .target()
        .ok_or_else(|| anyhow::anyhow!("No local target"))?;
    let base_oid = base_branch
        .target()
        .ok_or_else(|| anyhow::anyhow!("No base target"))?;

    let (ahead, _behind) = repo.graph_ahead_behind(local_oid, base_oid)?;
    Ok(ahead > 0)
}

/// Create RepoInfo for the manifest repository
fn create_manifest_repo_info(
    config: &crate::core::manifest::ManifestRepoConfig,
    _manifests_dir: &PathBuf,
    workspace_root: &PathBuf,
) -> Option<RepoInfo> {
    let path = ".gitgrip/manifests".to_string();

    crate::core::repo::RepoInfo::from_config(
        "manifest",
        &crate::core::manifest::RepoConfig {
            url: config.url.clone(),
            path,
            default_branch: config.default_branch.clone(),
            copyfile: config.copyfile.clone(),
            linkfile: config.linkfile.clone(),
            platform: config.platform.clone(),
            reference: false,
        },
        workspace_root,
    )
}

/// Check if a repo has changes ahead of its default branch
/// Returns Ok(true) if there are changes, Ok(false) if no changes or on default branch
fn check_repo_for_changes(
    repo: &RepoInfo,
    branch_name: &mut Option<String>,
) -> anyhow::Result<bool> {
    let git_repo = open_repo(&repo.absolute_path)
        .map_err(|e| anyhow::anyhow!("Failed to open repo: {}", e))?;

    let current = get_current_branch(&git_repo)
        .map_err(|e| anyhow::anyhow!("Failed to get current branch: {}", e))?;

    // Skip if on default branch
    if current == repo.default_branch {
        return Ok(false);
    }

    // Check for changes ahead of default branch
    let has_changes = has_commits_ahead(&git_repo, &current, &repo.default_branch)
        .map_err(|e| anyhow::anyhow!("Failed to check commits: {}", e))?;

    if has_changes {
        // Update branch_name for consistency checking
        if branch_name.is_none() {
            *branch_name = Some(current);
        }
        Ok(true)
    } else {
        Ok(false)
    }
}

/// Get authentication token for platform
#[allow(dead_code)]
pub fn get_token_for_platform(platform: &PlatformType) -> Option<String> {
    match platform {
        PlatformType::GitHub => std::env::var("GITHUB_TOKEN")
            .ok()
            .or_else(|| std::env::var("GH_TOKEN").ok()),
        PlatformType::GitLab => std::env::var("GITLAB_TOKEN").ok(),
        PlatformType::AzureDevOps => std::env::var("AZURE_DEVOPS_TOKEN").ok(),
    }
}
